subprojects/snapshots/src/main/java/org/gradle/internal/vfs/VirtualFileSystem.java

```diff
@@ -50,16 +50,16 @@ public interface VirtualFileSystem {
      * If the snapshotted location is invalidated while snapshotting,
      * then the snapshot is not stored in the VFS to avoid inconsistent state.
      */
-    FileSystemLocationSnapshot store(String baseLocation, StoringAction storingAction);
+    <T> T store(String baseLocation, StoringAction<T> storingAction);
 
     /**
      * Snapshotting action which produces possibly more than one snapshot.
      *
      * For example when snapshotting a filtered directory, the snapshots for complete subdirectories
      * would be reported here when they are found.
      */
-    interface StoringAction {
-        FileSystemLocationSnapshot snapshot(Consumer<FileSystemLocationSnapshot> snapshot);
+    interface StoringAction<T> {
+        T snapshot(Consumer<FileSystemLocationSnapshot> snapshot);
```

Comment:
I think using a specific interface here instead of `Consumer` could clarify the intent on the caller side. So instead of `vfsStorer.accept()` we could say `vfsStorer.store()`. WDYT?

Reply:
Actually, the mechanics here feel brittle. IIUC, the implementor of `StoringAction` is supposed to call `snapshot.accept()` and then return a potentially transformed value. I think this would be better if the storing would happen automatically.

One way to do this would be like this:

```java
<T, S extends FileSystemLocationSnapshot> T store(String baseLocation, Supplier<? extends S> snapshotter, Function<? super S, ? extends T> transformer);
```

Reply:
That doesn't work, since for directory snapshotting we will be storing things along the way and not the full snapshot we build up at the end. The storing action was supposed for this use-case. Maybe we'd need something like `maybeStore`?